mirgecom/diffusion.py

```diff
@@ -49,24 +53,102 @@ def _u_flux(discr, alpha, q_tpair):
     return discr.project(q_tpair.dd, "all_faces", flux_weak)
 
 
-def diffusion_operator(discr, alpha, u):
+class DiffusionBoundary(metaclass=abc.ABCMeta):
+    """
+    Diffusion boundary base class.
+
+    .. automethod:: get_q_flux
+    .. automethod:: get_u_flux
+    """
+
+    @abc.abstractmethod
+    def get_q_flux(self, discr, alpha, dd, u):
+        """Compute the flux for *q* on the boundary corresponding to *dd*."""
+        pass
+
+    def get_u_flux(self, discr, alpha, dd, q):
+        """Compute the flux for *u* on the boundary corresponding to *dd*."""
+        pass
+
+
+class DirichletDiffusionBoundary(DiffusionBoundary):
+    """
+    Dirichlet boundary condition for the diffusion operator.
+
+    .. automethod:: __init__
+    """
+
+    def __init__(self, value=0.):
+        """
+        Initialize the boundary condition.
+
+        Parameters
+        ----------
+        value: float or meshmode.dof_array.DOFArray
+            the value(s) along the boundary
+        """
+        self.value = value
+
+    def get_q_flux(self, discr, alpha, dd, u):  # noqa: D102
+        dir_u = discr.project("vol", dd, u)
+        u_tpair = TracePair(dd, interior=dir_u, exterior=2.*self.value-dir_u)
+        return _q_flux(discr, alpha, u_tpair)
+
+    def get_u_flux(self, discr, alpha, dd, q):  # noqa: D102
+        dir_q = discr.project("vol", dd, q)
+        q_tpair = TracePair(dd, interior=dir_q, exterior=dir_q)
+        return _u_flux(discr, alpha, q_tpair)
+
+
+class NeumannDiffusionBoundary(DiffusionBoundary):
+    """
+    Neumann boundary condition for the diffusion operator.
+
+    .. automethod:: __init__
+    """
+
+    def __init__(self, value=0.):
+        """
+        Initialize the boundary condition.
+
+        Parameters
+        ----------
+        value: float or meshmode.dof_array.DOFArray
+            the value(s) along the boundary
+        """
+        self.value = value
+
+    def get_q_flux(self, discr, alpha, dd, u):  # noqa: D102
+        dir_u = discr.project("vol", dd, u)
+        u_tpair = TracePair(dd, interior=dir_u, exterior=dir_u)
+        return _q_flux(discr, alpha, u_tpair)
+
+    def get_u_flux(self, discr, alpha, dd, q):  # noqa: D102
+        ones = discr.zeros(q[0].array_context) + 1.
+        dir_ones = discr.project("vol", dd, ones)
+        flux_weak = alpha*self.value*dir_ones
+        return discr.project(dd, "all_faces", flux_weak)
+
+
+def diffusion_operator(discr, alpha, boundaries, u):
     r"""
     Compute the diffusion operator.
 
     The diffusion operator is defined as
     $\nabla\cdot(\alpha\nabla u)$, where $\alpha$ is the diffusivity and
     $u$ is a scalar field.
 
-    Currently sets $u = 0$ on the boundaries.
-
     Parameters
     ----------
     discr: grudge.eager.EagerDGDiscretization
         the discretization to use
     alpha: float
         the (constant) diffusivity
+    boundaries:
+        dictionary (or object array of dictionaries) mapping boundary tags to
```

Comment:
Object array of dictionaries doesn't make much sense, since dictionaries don't meaningfully support arithmetic. Tuple or list?

Reply:
It's only an object array when `u` is an object array. I think I need that for `obj_array_vectorize_n_args` to work, right?

Reply:
If you'd like to use them internally to make use of `obj_array_vectorize_n_args`, that's OK, but IMO it's not a sane interface to expose to the user.

Reply:
(I.e. you could accept a list and internally convert it.)